src/exec/use_pty/pipe.rs

```diff
@@ -142,78 +156,202 @@ impl<R: Read, W: Write> Buffer<R, W> {
+            //         head         tail
+            // In the extreme case where tail is the last position, the sub-buffer will have a
+            // length of one.
+            //                             ┌───┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │   │ M │ D │ D │ D │ D │ D │   │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //           ▲                   ▲
+            //           │                   │
+            //         head                 tail
+            // We can be sure that the marker byte is not at tail because the buffer is not full.
+            // On the other hand if tail is equal to head, the buffer is empty
+            //                     ┌───────────┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │   │   │   │   │ M │   │   │   │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //                       ▲
+            //                       │
+            //                 head and tail
+            //
+            &mut self.buffer[self.tail..]
+        } else {
+            // If tail is less than head, the empty sub-buffer extends from tail to the marker byte
+            // without including the latter. We know that the marker byte is one position before
+            // head.
+            //     ┌───────────┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │ D │   │   │   │ M │ D │ D │ D │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //       ▲               ▲
+            //       │               │
+            //     tail             head
+            //
+            // In the extreme case where tail is as close as it can be to head, the sub-buffer will
+            // have a length of one. We can be sure that the marker byte is not at tail because the
+            // buffer is not full.
+            //     ┌───┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │ D │   │ M │ D │ D │ D │ D │ D │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //       ▲       ▲
+            //       │       │
+            //     tail     head
+            &mut self.buffer[self.tail..(self.head - 1)]
+        };
+
+        // Insert `len` bytes from `read` into the buffer.
         let len = read.read(buffer)?;
 
-        // Mark the `len` bytes after the busy section as busy too.
-        self.end += len;
-
-        // If we read something, the buffer is not empty anymore and we can resume writing.
         if len > 0 {
+            // We update tail so it becomes the next position where we can insert data again.
+            self.tail += len;
+            // However, we must be sure that tail doesn't go past the last position.
+            self.tail %= self.buffer.len();
+            // If we actually inserted something, the buffer is not empty anymore and we can resume
+            // writing.
             self.write_handle.resume(registry);
         }
```

Comment:
if `len == buffer.len()` then you could do another `read.read`. that condition indicates that you filled `buffer` but that does not necessarily mean that you have read all the available data out of `read`. it could also be the the case that there's no more space available in the circular buffer (e.g. in the `tail > head`) case.

I would suggest adding a helper method that returns the next chunk of available space from the ring buffer. it would probably also be clearer to refactor the ring buffer (the `buffer`, `head` and `tail` fields) into its own struct. in pseudo-code you may end with something like this

``` rust
let chunk = self.ring_buffer.next_writable_chunk();

if chunk.len() == 0 {
    // if the ring buffer is full you may want to signal something to your event loop (?)
    return;
}

let written_bytes = read.read(chunk);
let there_may_be_more_available_data = written_bytes == chunk.len();

self.ring_buffer.advance_tail(written_bytes);

if there_may_be_more_available_data {
    let chunk = self.ring_buffer.next_writable_chunk();

    // use `if` to avoid a no-op syscall
    if !chunk.is_empty() {
        let written_bytes = read.read(chunk);

        self.ring_buffer.advance_tail(written_bytes);
    }
}

// If we actually inserted something, the buffer is not empty anymore and we can resume
// writing.
self.write_handle.resume(registry);

```